### PR TITLE
implemente proposeblock handler and SyncInfo handler

### PR DESCRIPTION
**Pull request by wjrjerome**

### CODE REVIEW

File: consensus/XDPoS/engines/engine_v2/engine.go [hunk 1]
```diff
@@ -124,12 +126,16 @@ func (x *XDPoS_v2) VerifySyncInfoMessage(syncInfo *utils.SyncInfo) error {
...
-func (x *XDPoS_v2) SyncInfoHandler(syncInfo *utils.SyncInfo) error {
+func (x *XDPoS_v2) SyncInfoHandler(chain consensus.ChainReader, syncInfo *utils.SyncInfo) error {
 	/*
 		1. processQC
 		2. processTC
 	*/
-	return nil
+	err := x.processQC(chain, syncInfo.HighestQuorumCert)
+	if err != nil {
+		return nil
```
Comment (wgr523) [inline consensus/XDPoS/engines/engine_v2/engine.go]: return err?

File: consensus/XDPoS/engines/engine_v2/engine.go [hunk 2]
```diff
@@ -267,26 +273,38 @@ func (x *XDPoS_v2) onTimeoutPoolThresholdReached(pooledTimeouts map[common.Hash]
 }
 
 /*
-	Process Block workflow
+	Proposed Block workflow
```
Comment (wgr523) [inline consensus/XDPoS/engines/engine_v2/engine.go:278]: I suggest to add more description like this:
Proposed Block workflow:
1. process the QC inside the proposed block
2. the proposed block's info is extracted into BlockInfo and verified for voting

File: consensus/XDPoS/engines/engine_v2/engine.go [hunk 3]
```diff
@@ -124,12 +126,16 @@ func (x *XDPoS_v2) VerifySyncInfoMessage(syncInfo *utils.SyncInfo) error {
 	return nil
 }
 
-func (x *XDPoS_v2) SyncInfoHandler(syncInfo *utils.SyncInfo) error {
+func (x *XDPoS_v2) SyncInfoHandler(chain consensus.ChainReader, syncInfo *utils.SyncInfo) error {
```
Comment (wgr523) [inline consensus/XDPoS/engines/engine_v2/engine.go:129]: do we need lock here?

File: consensus/XDPoS/engines/engine_v2/engine.go [hunk 4]
```diff
@@ -376,22 +393,37 @@ func (x *XDPoS_v2) setNewRound(round utils.Round) error {
 }
 
 // Hot stuff rule to decide whether this node is eligible to vote for the received block
-func (x *XDPoS_v2) verifyVotingRule(header *types.Header) error {
+func (x *XDPoS_v2) verifyVotingRule(blockCahinReader consensus.ChainReader, blockInfo *utils.BlockInfo, quorumCert *utils.QuorumCert) (bool, error) {
```
Comment (wgr523) [inline consensus/XDPoS/engines/engine_v2/engine.go]: blockCahinReader -> blockChainReader

File: consensus/XDPoS/engines/engine_v2/engine.go [hunk 5]
```diff
@@ -376,22 +393,37 @@ func (x *XDPoS_v2) setNewRound(round utils.Round) error {
...
+	// Make sure this node has not voted for this round. We can have a variable highestVotedRound, and check currentRound > highestVotedRound.
```
Comment (wgr523) [inline consensus/XDPoS/engines/engine_v2/engine.go]: remove my comment, this sentence: ~~We can have a variable highestVotedRound, and check currentRound > highestVotedRound.~~

File: consensus/XDPoS/engines/engine_v2/engine.go [hunk 6]
```diff
@@ -401,6 +433,7 @@ func (x *XDPoS_v2) sendVote(blockInfo *utils.BlockInfo) error {
 		Signature:         signedHash,
 	}
 	x.broadcastToBftChannel(voteMsg)
+	x.highestVotedRound = x.currentRound
```
Comment (wgr523) [inline consensus/XDPoS/engines/engine_v2/engine.go]: I suggest make this the first step. It makes sure it sets highestVotedRound first under possible crashing.

File: consensus/XDPoS/engines/engine_v2/engine.go [hunk 7]
```diff
@@ -519,3 +552,22 @@ func (x *XDPoS_v2) commitBlocks(blockCahinReader consensus.ChainReader, proposed
 
 	return true, nil
 }
+
+func (x *XDPoS_v2) isExtendingFromAncestor(blockCahinReader consensus.ChainReader, currentBlock *utils.BlockInfo, ancestorBlock *utils.BlockInfo) (bool, error) {
```
Comment (wgr523) [inline consensus/XDPoS/engines/engine_v2/engine.go]: blockCahinReader -> blockChainReader

File: tests/consensus/proposed_block_test.go [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+	"github.com/XinFinOrg/XDPoSChain/consensus/XDPoS"
+	"github.com/XinFinOrg/XDPoSChain/consensus/XDPoS/utils"
+	"github.com/XinFinOrg/XDPoSChain/params"
+	"github.com/stretchr/testify/assert"
+)
+
+// ProposeBlock handler
+func TestProposedBlockMessageHandlerSuccessfullyGenerateVote(t *testing.T) {
+	blockchain, _, currentBlock, _ := PrepareXDCTestBlockChainForV2Engine(t, 11, params.TestXDPoSMockChainConfigWithV2Engine)
```
Comment (wgr523) [inline tests/consensus/proposed_block_test.go:15]: is currentBlock no.11 block? then extra in it points to no.10 block

**Review by wgr523 — APPROVED** [review]

File: consensus/XDPoS/engines/engine_v2/engine.go (hunk 6 above)
Comment (wjrjerome) [reply consensus/XDPoS/engines/engine_v2/engine.go]: This have to be after the error checking, just in case throw error at `signSignature`

**Pull request merged**